File: Difficulty/Hard/[65] Valid Number/solution.py
class Solution:
    def isNumber(self, s: str) -> bool:
        s = s.strip()
        if not s:
            return False
        if len(s) == 1:
            return s[0] >= '0' and s[0] <= '9'
        if s[0] == '-' or s[0] == '+':
            s = s[1:]
        try:
            e_pos = s.index('e')
            if e_pos == len(s) - 1 or e_pos == 0:
                return False
            else:
                n = s[e_pos + 1] == '-' or s[e_pos + 1] == '+'
                if e_pos + 1 == len(s) - 1 and n == 1:
                    return False
                else:
                    return self._checkNumber(s[:e_pos], False) \
                           and self._checkNumber(s[e_pos+1+n:], True)
        except ValueError:
            return self._checkNumber(s, False)

    def _checkNumber(self, s: str, dot_exist: bool) -> bool:
        if len(s) == 1 and s[0] == '.':
            return False
        for i in range(len(s)):
            if s[i] >= '0' and s[i] <= '9':
                continue
            elif s[i] == '.' and not dot_exist:
                dot_exist = True
            else:
                return False
        return True
        

File: Difficulty/Hard/[65] Valid Number/solution.cpp
class Solution {
public:
    bool isNumber(string s) {
        if (s.empty()) return 0;
        s.erase(0, s.find_first_not_of(' '));
        s.erase(s.find_last_not_of(' ')+1);
        if (s[0] == '-' || s[0] == '+')
            s.erase(0, 1);
        if (s.size() == 1) return s[0] >= '0' && s[0] <= '9';
        size_t e_pos = s.find_first_of('e');
        if (e_pos != string::npos && e_pos != s.size() - 1 && e_pos != 0){
            int n = s[e_pos + 1] == '-' || s[e_pos + 1] == '+';
            if (e_pos + 1 == s.size() - 1 && n == 1) return false;
            else return checkNumber(s.substr(0, e_pos), false) & checkNumber(s.substr(e_pos+1+n), true);
        }
        else if (e_pos == s.size() - 1 || e_pos == 0) 
            return false;
        else
            return checkNumber(move(s), false);
    }

    bool checkNumber(string&& s, bool dot_exist){
        if (s.size() == 1 && s[0] == '.') return false;
        for (int i = 0; i < s.size(); ++i){
            if (s[i] >= '0' && s[i] <= '9')
                continue;
            else if (s[i] == '.' && !dot_exist)
                dot_exist = true;
            else 
                return false;
        }
        return true;
    }
};